### PR TITLE
Preliminary support for Python 3.13a3

### PR DESCRIPTION
**Pull request by icemac**
Open issues
- [x] tests segfault in 3.13 – probably I introduced an error into the C code.

### CODE REVIEW

**Comment by icemac** [comment]
I have no idea what I made wrong here, so I'll stop working on this PR.

**Comment by davisagli** [comment]
@icemac I fixed the segfault. PyObject_GetOptionalAttr expects the attribute name as a Python string, not a _PY_IDENTIFIER. The fixed implementation is based on the implementation of _PyObject_LookupAttrId that was removed from CPython.

**Comment by icemac** [comment]
@davisagli Thank you for making the build green.

**Comment by icemac** [comment]
After this PR is merged and released https://github.com/zopefoundation/zope.container/pull/53 could turn green.

**Review by d-maurer — APPROVED** [review]

File: src/zodbpickle/_pickle_33.c [hunk 1]
```diff
@@ -23,8 +23,31 @@ _Py_SET_SIZE(PyVarObject *ob, Py_ssize_t size)
...
+ */
+static int _PyObject_HasAttrId(PyObject* obj, void* id)
+{
+    int result;
+    PyObject *oname = _PyUnicode_FromId(id); /* borrowed */
+    if (!oname) {
+        return -1;
+    }
+    PyObject* attr_val;
+    result = PyObject_GetOptionalAttr(obj, oname, &attr_val);
```
[the remark this answers is not in the document]
Comment (icemac) [reply src/zodbpickle/_pickle_33.c:40]: @davisagli What do you think about this suggestion?

**Comment by d-maurer** [comment]
David Glick wrote at 2024-2-14 00:04 -0800:
> ...
>@d-maurer @icemac I don't know the consequences of ignoring all exceptions vs only AttributeError, so I'd prefer to keep the implementation as it is.

In general, it is a bad idea to use `hasattr` (i.e. to ignore all
exceptions) on a persistent object:
the access could cause a `ConflictError` which `hasattr` would turn
into `False` even though the object might have that attribute but the
object is not loadable at the moment. Instead of returning `False`,
the request should get retried in that case.

I assume that even though `zodbpickle` works with persistent objects,
the function at hand will be called only for special
methods (`__getstate__`, `__setstate__`, ...) which are not called
on the object itself but on its class.
Therefore, `hasattr` might be safe, but I do not mind to
retain the current implementation.

**Comment by icemac** [comment]
Thank you for reviewing resp. working on this PR. 😃

**Pull request merged**